src/idlcxx/src/streamer_generator.c

```diff
@@ -532,6 +547,27 @@ void load_locals(context_t* ctx, bool* storage)
   ctx->key_funcs.sequenceentriespresent = storage[3];
 }
 
+bool has_keys(idl_type_spec_t* spec)
+{
+  //is struct
+  if (idl_is_struct(spec))
+  {
+    idl_member_t* mem = ((idl_struct_t*)spec)->members;
+    while (mem)
+    {
+      if (idl_is_masked(mem, IDL_KEY)) return true;
```

Comment:
The type_spec for a member can be a struct itself, shouldn't this be recursive? Actually, I'd probably unalias the type_spec first and then see if it's a struct (or a union?).

Reply:
That is not done here, because it is not necessary.
This function checks if any of the members are set to be key values.
This is used to generate the correct key stream for structs and unions.
Only if none of the members are set as keys, should the complete struct be added to the key stream, if it itself is set to be a key value.
If any of the member struct's members are set as keys, then only the members that are set to key are added to the key stream.

struct a {
  long b_1;
  long b_2;
};

struct c {
  @key a m_a;
};

struct d {
  @key long e_1;
  long e_2;
};

struct f {
  @key d m_d;
};

In the case of c, the key stream will contain both b_1 and b_2, while in the case of f, the key stream will only contain e_1.